libs/src/sdk/services/WalletApi/DelegatedWalletApi.ts

```diff
@@ -0,0 +1,55 @@
+import { signTypedData } from 'eth-sig-util'
+import * as secp from '@noble/secp256k1'
+import type { UsersApi } from '../../api/generated/default'
+import type { WalletApiService } from './types'
+
+export class DelegatedWalletApi implements WalletApiService {
+  private apiKey: string
+  private apiSecret: string
+  private users?: UsersApi
+
+  constructor(apiKey: string, apiSecret: string) {
+    this.apiKey = apiKey
+    this.apiSecret = apiSecret
+  }
+
+  setUsersApi(users: UsersApi) {
+    this.users = users
+  }
+
+  getSharedSecret: (publicKey: string | Uint8Array) => Promise<Uint8Array> =
+    () => {
+      throw new Error('WalletApi not initialized')
+    }
+
+  sign = async (data: string, userId?: string) => {
```

Comment:
Randomly saw this, sorry to poke my head in 😂 

Worried slightly about circular deps etc w.r.t. the dep on UsersApi here. This sort of breaks the intent of the "services layer" being isolated from the "API layer", and I imagine at some point the UsersApi will have a dep on WalletApiService. Also having that typecast in sdk is not great...

Curious if it'd be too annoying to plumb the wallet (pubkey) directly from the consumer instead of just the userId? Or failing that, if TracksApi can dep on UsersApi instead?

It might make the interface slightly more cumbersome, though, which is not ideal... curious on your thoughts. Trying to avoid having too much interdependency with walletApi as I think that's the most confusing part of the existing libs arch (the Web3Manager)

Reply:
No need to apologize! We had a feeling this was a bit jank too, but had trouble thinking of something better given the user pubkey requirement.
I think what we have is better than getting the pubkey directly from the consumer since ideally the wallet concept is always abstracted away for our end users. IMO would be better to keep the interface as simple as possible even if it makes the SDK code more complicated.
Maybe something in the middle could be: we have the consumer do something like this:
```
audiusSdk.initUser('h33j'); // name TBD but this must be called once before any delegated writes for a given user. Under the hood, this would just fetch and cache the user's pub key.
audiusSdk.uploadTrack({owner_id: 'h33j', ...etc})
```
Of course this would still be less ideal for the consumer.

We can def have TracksApi dep on UsersApi instead but then we'd have to add the same pubkey-fetching logic to all of the write paths/Apis. I guess that might not be so bad though with some good decomping.  Also, we were thinking of keeping a record of user ids to pubkeys in WalletAPI to avoid repeat pubkey fetching - I think we could still do this either way.

Down for us 3 to huddle about this!

Reply:
@sliptype / @nicoback2 I can huddle after client standup if you're free?

My latest proposal idea
- Have DelegatedWalletApi take a GeneratedUsersApi in the constructor
- When we make the UsersApi (with the wallet API and others baked in) that will be a separate instance
- DelegatedWalletApi caches UserId => Pubkey map

Definitely feel the need to keep this dead simple and lightweight, especially for the mass uploading usecase, and appreciate that there’s a tough balance to strike. I think this gives us an out from the potential circular dep, and also from the jank cast + setter?

Alternatively, you _could_ have `uploadTrack` accept a full user metadata (leaving the "wallet" part still somewhat abstracted) but you're right that's a lot more work for the end consumer.

Reply:
@rickyrombo / @nicoback2 What do yall think about this? https://github.com/AudiusProject/audius-protocol/pull/5232/commits/a4a3bc02ba8f8bc9c891d8e815c740b2ee42c837

Instead of making DelegatedWalletApi dependent on UsersApi, I added another argument to `manageEntity` (we were already passing WalletApi in here). Then I pass the UsersApi to the TracksApi constructor, now `uploadTrack` is responsible for passing UsersApi into `manageEntity`. `manageEntity` fetches the wallet and passes it into `signTransaction`.

This makes it so the service constructors aren't dependent on the api layer. Also I don't foresee places other than `manageEntity` calling `signTransaction` so I think the fetch for user wallet is ok to have there

Also I disambiguated `sign` and `signTransaction` in walletApi because they are being used for different purposes and have different function signatures

Reply:
@sliptype / @nicoback2 I *think* I might have a weird but happy solution - can we sync in like 30min? Tl;dr: Make the whole generated `readApi` _also_ a service.